### PR TITLE
Handle concurrent dns_record integrations

### PR DESCRIPTION
**Pull request by nrobinaubertin**
Applicable spec: <link>

### Overview

This PR introduce changes allowing the bind-operator charm to relate with multiple charms through the `dns_record` relation.  
It does not try to handle conflicts for now, this will be for a subsequent PR.  
The integration tests are adapted to test multiple concurrent relations.

### Library Changes

I changed the library to return the relation data of all relations through the DNSRecordsProvider. I decided to make this only a patch-level change since we're the only ones using the library for now and it is in its infancy. I'm aware that this should be counted as a major version change later on.

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: lib/charms/bind/v0/dns_record.py [hunk 1]
```diff
@@ -623,14 +624,16 @@ def __init__(self, charm: ops.CharmBase, relation_name: str = DEFAULT_RELATION_N
 
     def get_remote_relation_data(
         self,
-    ) -> Optional[Tuple[DNSRecordRequirerData, DNSRecordProviderData]]:
-        """Retrieve the remote relation data.
+    ) -> List[Optional[Tuple[DNSRecordRequirerData, DNSRecordProviderData]]]:
```
Comment (arturo-seijas) [inline lib/charms/bind/v0/dns_record.py:627]: Doesn't 'List` instead of  `List[Optional]` suffice?

File: src/bind.py [hunk 2]
```diff
@@ -184,30 +184,49 @@ def _generate_named_conf_local(self, zones: typing.List[str]) -> str:
...
             A resulting DNSRecordProviderData to put in the relation databag
         """
-        zones = self._to_bind_zones(rrd)
-        logger.debug("ZONES: %s", zones)
+        # Create staging area
         tempdir = tempfile.mkdtemp()
+
+        # Write zone files
+        zones = {}
+        for rrd, _ in relation_data:
```
Comment (arturo-seijas) [inline src/bind.py:204]: rrd?

File: lib/charms/bind/v0/dns_record.py (hunk 1 above)
Comment (nrobinaubertin) [reply lib/charms/bind/v0/dns_record.py]: You're right

File: src/bind.py (hunk 2 above)
Comment (nrobinaubertin) [reply src/bind.py]: that was for record_requirer_data, I replaced it

**Review by arturo-seijas — APPROVED** [review]

**Review by amandahla — APPROVED** [review]
LGTM

**Pull request merged**